### PR TITLE
fix: primary sp should check received ssp signature before sealing.

### PR DESCRIPTION
**Pull request by alexgao001**
### Description

primary sp should check received ssp bls signature before sealing object. If the bls signature is invalid. the secondary sp will be added to freeze pool as well.

### Rationale

tell us why we need these changes...

### Example

add an example CLI or API response...

### Changes

Notable changes: 
* add each change in a bullet point here
* ...

### CODE REVIEW

File: modular/executor/execute_replicate.go [hunk 1]
```diff
@@ -122,13 +125,26 @@ func (e *ExecuteModular) handleReplicatePiece(ctx context.Context, rTask coretas
...
+		}
+		for rIdx, spEp := range rTask.GetSecondaryEndpoints() {
+			log.Debugw("start to done replicate", "sp", spEp)
+			signature, innerErr := e.doneReplicatePiece(ctx, rTask, spEp, int32(rIdx))
 			if innerErr == nil {
+				msg := storagetypes.NewSecondarySpSealObjectSignDoc(e.baseApp.ChainID(), gvg.Id, rTask.GetObjectInfo().Id, storagetypes.GenerateHash(rTask.GetObjectInfo().GetChecksums()[:])).GetBlsSignHash()
+				err = veritySecondarySpBlsSignature(e.getSpByID(gvg.GetSecondarySpIds()[rIdx]), signature, msg[:])
+				if err != nil {
+					rTask.SetNotAvailableSpIdx(int32(rIdx))
+					log.CtxErrorw(ctx, "verify secondary SP bls signature failed", "SP id:", gvg.GetSecondarySpIds()[rIdx], "err:", err.Error())
```
Comment (will-2012) [inline modular/executor/execute_replicate.go:141]: `SP id:`-->`secondary_sp_id`
`err:`-->`error`

File: modular/executor/executor.go [hunk 2]
```diff
@@ -83,16 +96,33 @@ func (e *ExecuteModular) eventLoop(ctx context.Context) {
...
+			}
+			e.mutex.RUnlock()
 		}
 	}
 }
 
+func (e *ExecuteModular) getSpByID(id uint32) *sptypes.StorageProvider {
+	e.mutex.RLock()
+	defer e.mutex.RUnlock()
+	return e.spMap[id]
```
Comment (will-2012) [inline modular/executor/executor.go:123]: Maybe return nil and panic.
Refer to SPCachePool?

Comment (alexgao001) [reply modular/executor/executor.go:123]: There is no eviction so shouldn't be `nil`

File: modular/executor/execute_replicate.go (hunk 1 above)
Comment (alexgao001) [reply modular/executor/execute_replicate.go:141]: fixed

File: modular/executor/executor.go [hunk 3]
```diff
@@ -83,16 +96,33 @@ func (e *ExecuteModular) eventLoop(ctx context.Context) {
...
 		case <-ctx.Done():
 			return
 		case <-statisticsTicker.C:
 			log.CtxInfo(ctx, e.Statistics())
+		case <-updateSpTicker.C:
+			sps, err := e.baseApp.Consensus().ListSPs(ctx)
+			if err != nil {
+				continue
+			}
+			e.mutex.RLock()
```
Comment (will-2012) [inline modular/executor/executor.go:111]: WLock?

Comment (alexgao001) [reply modular/executor/executor.go]: Yes.

File: modular/executor/execute_replicate.go [hunk 4]
```diff
@@ -122,13 +125,26 @@ func (e *ExecuteModular) handleReplicatePiece(ctx context.Context, rTask coretas
...
+		}
+		for rIdx, spEp := range rTask.GetSecondaryEndpoints() {
+			log.Debugw("start to done replicate", "sp", spEp)
+			signature, innerErr := e.doneReplicatePiece(ctx, rTask, spEp, int32(rIdx))
 			if innerErr == nil {
+				msg := storagetypes.NewSecondarySpSealObjectSignDoc(e.baseApp.ChainID(), gvg.Id, rTask.GetObjectInfo().Id, storagetypes.GenerateHash(rTask.GetObjectInfo().GetChecksums()[:])).GetBlsSignHash()
+				err = veritySecondarySpBlsSignature(e.getSpByID(gvg.GetSecondarySpIds()[rIdx]), signature, msg[:])
+				if err != nil {
+					rTask.SetNotAvailableSpIdx(int32(rIdx))
+					log.CtxErrorw(ctx, "verify secondary SP bls signature failed", "secondary_sp_id", gvg.GetSecondarySpIds()[rIdx], "error", err.Error())
```
Comment (sysvm) [inline modular/executor/execute_replicate.go:141]: fix as `failed to verify xxx` is better.

File: modular/executor/execute_replicate.go [hunk 5]
```diff
@@ -6,6 +6,9 @@ import (
 	"sync"
 	"time"
 
+	sptypes "github.com/bnb-chain/greenfield/x/sp/types"
+	virtualgrouptypes "github.com/bnb-chain/greenfield/x/virtualgroup/types"
```
Comment (sysvm) [inline modular/executor/execute_replicate.go:10]: sort imported pkg.

File: modular/executor/execute_replicate.go [hunk 6]
```diff
@@ -316,19 +332,17 @@ func (e *ExecuteModular) doneReplicatePiece(ctx context.Context, rTask coretask.
...
 	}
-	originMsgHash := storagetypes.NewSecondarySpSealObjectSignDoc(sdkmath.NewUint(objectID), gvgId, integrity).GetBlsSignHash()
-	err := types.VerifyBlsSignature(blsPubKey, originMsgHash, signature)
+	sig, err := bls.SignatureFromBytes(signature)
 	if err != nil {
-		log.CtxErrorw(ctx, "failed to verify signature", "error", err)
 		return err
 	}
+	if !sig.Verify(publicKey, sigDoc) {
+		return fmt.Errorf("sp %d bls signature verification failed", secondarySp.Id)
```
Comment (sysvm) [inline modular/executor/execute_replicate.go:345]: ditto.

**Review by will-2012 — APPROVED** [review]

File: modular/executor/execute_replicate.go (hunk 5 above)
Comment (alexgao001) [reply modular/executor/execute_replicate.go:10]: How to sort it?

File: modular/executor/execute_replicate.go (hunk 6 above)
Comment (alexgao001) [reply modular/executor/execute_replicate.go]: fixed

File: modular/executor/execute_replicate.go (hunk 4 above)
Comment (alexgao001) [reply modular/executor/execute_replicate.go]: fixed

**Review by sysvm — APPROVED** [review]
LGTM

File: modular/executor/execute_replicate.go (hunk 5 above)
Comment (sysvm) [reply modular/executor/execute_replicate.go:10]: group it with bnb-chain library

**Pull request merged**